apps/aecore/lib/aecore/chain/chainstate.ex

```diff
@@ -104,7 +110,24 @@ defmodule Aecore.Chain.Chainstate do
   """
   @spec calculate_root_hash(t()) :: binary()
   def calculate_root_hash(chainstate) do
-    AccountStateTree.root_hash(chainstate.accounts)
+    [
+      AccountStateTree.root_hash(chainstate.accounts),
+      NamingStateTree.root_hash(chainstate.naming),
+      OracleStateTree.root_hash(chainstate.oracles)
+    ]
+    |> Enum.reduce(<<@protocol_version::size(@protocol_version_field_size)>>, fn root_hash, acc ->
+      acc <> pad_empty(root_hash)
+    end)
+    |> Hash.hash_blake2b()
+  end
+
+  defp pad_empty(root_hash_binary)
+       when is_binary(root_hash_binary) and byte_size(root_hash_binary) === @state_hash_bytes do
+    root_hash_binary
+  end
+
+  defp pad_empty(_) do
```

Comment:
When do we return zeros? (What's the case?)
I think we should adjust the checks so we crash when the provided hash has different length, not return zeros.

Reply:
You may be right @cytadela8 . I share a similar view with you. But that's the way they do it in the `epoch` and we just follow them.
https://github.com/aeternity/epoch/blob/master/apps/aecore/src/aec_trees.erl#L327-L328

Reply:
@cytadela8 
What I found is the following
There is a difference in the implementation of patricia merkle trie (PMT) library for erlang and for elixir

`epoch`
When creating a PMT, the library returns an `empty binary  = <<>>` for root_hash.
Then after, outside the PTM library, a check is made to determine whether the root_hash is empty or not.
And if it is `empty = <<>>`, the `{error, empty}` tuple is returned, otherwise `{ok, Hash}`. The both cases are then handled differently.

`elixir`
However, what is pledged as an elixir implementation is slightly different.
It is different from the moment the root_hash is generated.
There is a function that is called immediately after the creation of the PMT. The function is part of the library.
And again, a check is made to determine whether the root_hash is empty or not.
And if it is empty = <<>>, a second internal functionality is invoked which digest the empty binary and as a result returns `canonical binary`.

What we could do is to use this `canonical hash` as a `{error, empty}` to fit the scenario.

Reply:
> What we could do is to use this canonical hash as a {error, empty} to fit the scenario.
I think you should do so

Reply:
@thepiwo 
Done